src/transformers/pipelines/base.py

```diff
@@ -739,65 +796,72 @@ def check_model_type(self, supported_models: Union[List[str], dict]):
+        raise NotImplementedError("preprocess not implemented")
 
-        Args:
-            inputs: dict holding all the keyword arguments for required by the model forward method.
-            return_tensors: Whether to return native framework (pt/tf) tensors rather than numpy array
+    @abstractmethod
+    def postprocess(self, model_outputs: ModelOutput, **postprocess_parameters) -> Any:
+        raise NotImplementedError("postprocess not implemented")
 
-        Returns:
-            Numpy array
-        """
-        # Encode for forward
+    @abstractmethod
+    def forward(self, input_tensors: Dict[str, GenericTensor], **forward_parameters) -> ModelOutput:
+        raise NotImplementedError("postprocess not implemented")
+
+    def infer_forward(self, model_inputs, *args, **kwargs):
         with self.device_placement():
             if self.framework == "tf":
-                # TODO trace model
-                predictions = self.model(inputs.data, training=False)[0]
-            else:
+                model_inputs["training"] = False
+                model_outputs = self.forward(model_inputs)
+            elif self.framework == "pt":
                 with torch.no_grad():
-                    inputs = self.ensure_tensor_on_device(**inputs)
-                    predictions = self.model(**inputs)[0].cpu()
+                    model_inputs = self.ensure_tensor_on_device(**model_inputs)
+                    model_outputs = self.forward(model_inputs)
+            else:
+                raise ValueError(f"Framework {self.framework} is not supported")
+        return model_outputs
 
-        if return_tensors:
-            return predictions
+    def get_iterator(
+        self,
+        inputs,
+        num_workers,
+    ):
+        os.environ["TOKENIZERS_PARALLELISM"] = "false"
+        dataset = PipelineDataset(inputs, self.preprocess)
+        dataloader = DataLoader(dataset, num_workers=num_workers, batch_size=1, collate_fn=collate_fn)
+        model_iterator = PipelineIterator(dataloader, self.infer_forward)
+        final_iterator = PipelineIterator(model_iterator, self.postprocess)
+        return final_iterator
+
+    def __call__(self, inputs, *args, num_workers=8, **kwargs):
+        self.set_parameters(**kwargs)
+        self.call_count += 1
+        if self.call_count > 10 and self.framework == "pt" and self.device.type == "cuda":
+            warnings.warn(
+                "You seem to be using the pipelines sequentially on GPU. In order to maximize efficiency please use a dataset",
+                UserWarning,
+            )
+        if isinstance(inputs, list):
+            if self.framework == "pt":
+                final_iterator = self.get_iterator(inputs, num_workers)
+                outputs = [output for output in final_iterator]
+                return outputs
+            else:
+                return self.run_multi(inputs)
+        elif Dataset and isinstance(inputs, Dataset):
```

Comment:
```suggestion
        elif Dataset is not None and isinstance(inputs, Dataset):
```
We always use explicit statements in the Transformers library.

Reply:
Agree!

Reply:
Done.